### DISCUSSION TITLE
wie Brennerstörung im Programm erkennen?!?

### DISCUSSION THREAD

**Discussion by MisterButcher** [dewenni/ESP_Buderus_KM271#103] [Q&A] [answered]
Hallo, 
ich habe die Softwareversion 3.2.4 als Basis für mein eigenes Heizungsmonitoring verwendet. Mir geht es primär darum eine Brennerstörung zu erkennen, und dann entsprechende Aktionen auszuführen (z.B. einen Pushbutton aktivieren der den Entstörknopf am Brenner drückt). Ich verwende also kein MQTT um die Daten zu sammeln und dann auszuwerten.
Wie auch immer, ich sehe wie die Kommunikation mit dem KM271 <-> MQTT durchgeführt wird, aber es fällt mir schwer die Stelle zu finden wo ich "brennerstoerung=true" einbauen könnte.

Kann mir jemand weiterhelfen wie ich im Programm / per Programm erkenne, dass eine Brennerstörung vorliegt?

Danke + Gruss

Stefan

**Comment by dewenni** [comment]
Hi,
Ich kann gerne helfen deine Anforderung umzusetzen, allerdings müsstest du hierzu noch ein paar mehr Infos geben.

Wie würdest du den die Aktion zum Brenner Reset umsetzten wollen? Würdest dazu einen Kanal vom Board als Ausgang nutzen wollen?

Und dieser Ausgang soll dann von der Software getriggert werden wenn eine Brennerstörung vorliegt?

Das wäre dann nur eine spezielle Lösung für dich und du müsstest das dann für dich im Code umsetzten und könntest dann nur schwierig auf neue Versionen Updaten.

Einfacher und flexibler wäre aus meiner Sicht, wenn du die bereits existierenden MQTT Nachrichten nutzen würdest und dann in einer externen Software wie Home Assistant, ioBroker, Node-Red oder ähnlichem darauf reagierst und einen entsprechenden Aktor ansteuerst.

Kommt aber immer darauf an was du hier ggf. schon im Einsatz hast.

Grüße Sven

**Comment by MisterButcher** [comment]
Hallo Sven, 
ja, mir geht es um eine Speziallösung. Wie geschrieben, ich will keine Heizungsstatistik fahren. Ich möchte nur sicherstellen, dass nach einer Brennerstörung entstört wird (in Maßen natürlich, mit Limits etc).

bisherige Lösung:

Vor dem Display des Steuergeräts steht eine Webcam. Die macht alle 10 Minuten ein Bild und überträgt es per ftp an meinen Server. Dort äuft ein Script, welches das Bild analysiert und erkennt, wenn es nicht "Sommer" oder "Automatik" ist (-> "Brenner"). Dann wird per http request der Alarmausgang der Webcam getriggert (die hat sowas, da liegen dann 5V an). Dieser Alarmausgang aktiviert ein Relais welches dann einen Schütz (ein etwas stärkeres Relais) aktiviert der dann den Entstörknopf am Brenner drückt. Dadurch wird die Heizung dann entstört.

Wenn ich das per ESP / KM271 mache, dann spare ich mir die Kamera, das Script, das damit verbunde notwendige Licht und das permanente nachjustieren, wenn mal irgendwas gewackelt hat (sich Licht und Kameraposition verändert haben).
Ich habe bereits alles (also die Scriptlogic) auf den ESP übertragen, die Kommunikation mit dem KM271 habe ich natürlich drin gelassen. 
Per Board-Ausgang kann ich wieder das Relais triggern und damit den Schütz und die Heizung entstören.

Jetzt fehlt mir - wie geschrieben - nur noch die Stelle wo ich meinem Programmteil sage, dass eine Brennerstörung vorliegt, also eine Variable auf "true" setze.

Mehr will ich nicht, auch keine Softwareupdate etc. Wenn diese Lösung so funktioniert, und ich die Kamera abbauen kann, dann bin ich glücklich und zufrieden.

Gruss
Stefan

**Comment by dewenni** [comment, marked as answer]
ok, das war mir nach dem ersten Post noch nicht so klar.

Die Information kannst du denke ich aus der Variablen `kmStatus.BoilerErrorStates` rausziehen. Diese Variable ist bitcodiert und wie folgt definiert:

Bit1 = "Brennerstoerung"
Bit2 = "Kesselfuehler"
Bit3 = "Zusatzfuehler"
Bit4 = "Kessel bleibt kalt"
Bit5 = "Abgasfuehler"
Bit6 = "Abgas ueber Grenzwert"
Bit7 = "Sicherungskette ausgeloest"
Bit8 = "Externe Stoerung"

So könntest du darauf zugreifen:

```c++
#include <km271.h>

s_km271_status *pkmStatus = km271GetStatusValueAdr();

void checkError() {

  if (bitRead(pkmStatus->BoilerErrorStates, 0)) {
    // Fehler Brennerstoerung
  }
}
```

**Comment by MisterButcher** [comment]
Hallo Sven, 
danke, das hat mir sehr geholfen.
Ich habe es direkt in die Routine parseInfo in km271.cpp eingebaut:

```
case 0x8830: // 0x8830 : Bitfield
  kmStatus.BoilerErrorStates  = data[2];      

  // SRA085 
  if (bitRead(kmStatus.BoilerErrorStates, 0)) {
    sra_brennerstoerung = true;
    sra_logdata("ERR", "Brennerstoerung in BoilerErrorStates erkannt");
  }                         

  mqttPublish(addStatTopic(statTopic.BOILER_ERR_BURNER[config.lang]),      uint8ToString(bitRead(kmStatus.BoilerErrorStates, 0)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_SENSOR[config.lang]),      uint8ToString(bitRead(kmStatus.BoilerErrorStates, 1)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_AUX_SENS[config.lang]),    uint8ToString(bitRead(kmStatus.BoilerErrorStates, 2)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_STAY_COLD[config.lang]),   uint8ToString(bitRead(kmStatus.BoilerErrorStates, 3)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_GAS_SENS[config.lang]),    uint8ToString(bitRead(kmStatus.BoilerErrorStates, 4)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_EXHAUST[config.lang]),     uint8ToString(bitRead(kmStatus.BoilerErrorStates, 5)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_SAFETY[config.lang]),      uint8ToString(bitRead(kmStatus.BoilerErrorStates, 6)), false);
  mqttPublish(addStatTopic(statTopic.BOILER_ERR_EXT[config.lang]),         uint8ToString(bitRead(kmStatus.BoilerErrorStates, 7)), false);
  break;
```

Gruss
Stefan
